### ISSUE TITLE
Losing DDS connection with remote hosts

### ISSUE THREAD

**Issue by omertal88** [eclipse-cyclonedds/cyclonedds#1648]
Hello there,

I'm having a bit of a weird issue and I hope this thread might give me a lead on how to solve it.
**The scenario:**
I have two devices - both communicate through a wireless channel. One is a robot, and the other is the control station. Both running ROS2 galactic. I configured the discovery profiles to be such that the robot discovers the node on the control station, but not the other way around (I used simple discovery and specified the specfic port of the other node).
Sometimes (to the best of my knowledge) network connectivity becomes unstable (maybe due to radio noise - maybe not), and the node on the robot side will not receive messages on a topic sent by the control station.
I know for a fact that the node is publishing messages because I can see them using `ros2 topic echo`, and I know that the subscriber is listening, because when I do `ros2 topic pub`, and publish the message from CLI, I can see the node responding to the messages.

So if feels like CycloneDDS is losing the peer, but never regains connection (sorry if that's in layman's terms).
I looked in wireshark when this occurs and sure enough, I don't see the messages sent to the other node (I could only see METATRAFFIC sent from the node's source port).
I can, however, see discovery packets sent from the robot's node to the control station (so I would assume that user traffic should flow).

I'm attaching my wireshark captures from before and after resetting the control station node (which fixed the issue), but first I'll need to explain a few things:
1. The control station node is called `xtra_heartbeat_beacon` and is using participant index **8**.
2. The robot node is using a micro-ROS-Agent node (using 4 ports 7448-7451 for both meta and user traffic).
3. The micro ros agent is occupying the following ports:
```udp    UNCONN   0        0               127.0.0.1:41540          0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=27))                                     
udp    UNCONN   0        0                 0.0.0.0:49764          0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=26))                                     
udp    UNCONN   0        0               127.0.0.1:45830          0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=16))                                     
udp    UNCONN   0        0                 0.0.0.0:35994          0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=15))                                     
udp    UNCONN   0        0                 0.0.0.0:7448           0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=13))                                     
udp    UNCONN   0        0                 0.0.0.0:7449           0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=14))                                     
udp    UNCONN   0        0                 0.0.0.0:7450           0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=24))                                     
udp    UNCONN   0        0                 0.0.0.0:7451           0.0.0.0:*      users:(("micro_ros_agent",pid=4310,fd=25))                                     
```
4. The `xtra_heartbeat_beacon` node only communicates with the micro-ros-agent node.
5. There are other nodes running in the system, so probably best to filter the wireshark captures using the ports.
6. This issue occurs in all kinds of forms in other nodes as well, but since this node (`xtra_heartbeat_beacon`) is by far the simplest of all, I figured it's best to explain the issue using this node.
7. I also tried creating a node on the robot which uses cyclonedds to act as a bridge between `micro-ros-agent` and the `xtra_heartbeat_beacon` (just forwarding messages from one to another) so that it will be cyclonedds to cyclonedds, and the issue remains the same.
8. Before the reset `xtra_heartbeat_beacon` used `36038` as UDP output port, and after the reset - `53255`.
9. The topic name is `/ground_station/heartbeat` and the data is a simple string `gcs1`.

Before the reset:
[robot.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/11188937/robot.zip)
You can use this filter: `udp.srcport == 36038` to see the packets sent from `xtra_heartbeat_beacon` (again - should only be sent to ports 7448-7451).

After the reset:
[control.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/11188943/control.zip)
Here you can filter by: `udp.srcport == 53255`

I will edit this thread once I have cyclone logs and add them.

Thank you for any help :)
It will be most greatly appreciated.

**Edit:**
Okay so I was able to reproduce it by taking walking away far enough (until connection was lost) and head back. Then I saw that the robot doesn't receive messages from `xtra_heartbeat_beacon`. I have wireshark capture and cyclone logs (fine verbosity).
Hope this help
[cyclonedds-debug_xtra-heartbeat-beacon.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/11189799/cyclonedds-debug_xtra-heartbeat-beacon.zip)
[gcu-2023_04_10_13_01_35_filtered.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/11189931/gcu-2023_04_10_13_01_35_filtered.zip)
I had to crop the capture file to fit 25 MB - if you're missing something, I will find a way to upload the full one (36 MB).

`xtra_heartbeat_beacon` uses output port: `32820`.
micro-ros-agent uses:
```udp    UNCONN   0        0                 0.0.0.0:56043          0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=26))                                     
udp    UNCONN   0        0                 0.0.0.0:7448           0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=13))                                     
udp    UNCONN   0        0                 0.0.0.0:7449           0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=14))                                     
udp    UNCONN   0        0                 0.0.0.0:7450           0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=24))                                     
udp    UNCONN   0        0                 0.0.0.0:7451           0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=25))                                     
udp    UNCONN   0        0               127.0.0.1:49674          0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=16))                                     
udp    UNCONN   0        0               127.0.0.1:38008          0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=27))                                     
udp    UNCONN   0        0                 0.0.0.0:34600          0.0.0.0:*      users:(("micro_ros_agent",pid=4308,fd=15))                                     
```
By applying this filter: `ip.dst == 192.0.23.35 && (udp.srcport == 32820 && (udp.dstport == 7449 || udp.dstport == 7451))` I could see all user data packets sent from `xtra_heartbeat_beacon` to the `micro-ros-agent`. You can clearly see the traffic stops at some point (I've noticed that frame 32592 was the last message sent).

Also, this is the profile I used for the `xtra_heartbeat_beacon`:
```<?xml version="1.0" encoding="UTF-8" ?>
<CycloneDDS xmlns="https://cdds.io/config" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="https://cdds.io/config https://raw.githubusercontent.com/eclipse-cyclonedds/cyclonedd$
    <Domain id="0">
        <Compatibility>
            <ManySocketsMode>single</ManySocketsMode>
        </Compatibility>
        <Internal>
            <HeartbeatInterval max="inf" min="1 s" minsched="3 s"/>
            <SquashParticipants>true</SquashParticipants>
        </Internal>
        <General>
            <NetworkInterfaceAddress>eth0</NetworkInterfaceAddress>
            <AllowMulticast>false</AllowMulticast>
            <MaxMessageSize>65500 B</MaxMessageSize>
            <FragmentSize>1450B</FragmentSize>
            <MaxRexmitMessageSize>65500 B</MaxRexmitMessageSize>
            <Transport>udp</Transport>
        </General>
        <Discovery>
            <Peers>
<!-- Simple Peers -->

<!-- List of Local peers -->

<!-- List of Remote peers -->
            </Peers>
            <ParticipantIndex>8</ParticipantIndex>
            <MaxAutoParticipantIndex>50</MaxAutoParticipantIndex>
        </Discovery>
        <Tracing>
            <Verbosity>fine</Verbosity>
            <OutputFile>/opt/xtra/storage/debug/discovery/cyclonedds-debug_xtra-heartbeat-beacon.log</OutputFile>
        </Tracing>
    </Domain>
</CycloneDDS>
```

**Edit2:**
I was now able to reproduce with finest verbosity:
[gcu_2023_04_10_16_13.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/11190898/gcu_2023_04_10_16_13.zip)
[cyclonedds-debug_xtra-heartbeat-beacon_finest.zip](https://github.com/eclipse-cyclonedds/cyclonedds/files/11190900/cyclonedds-debug_xtra-heartbeat-beacon_finest.zip)
In this run:
```
user@gcu9a7f7611: ~$ sudo ss -tulpn | grep heartbeat
udp   UNCONN   0        0                  0.0.0.0:7426           0.0.0.0:*      users:(("xtra_heartbeat_",pid=21390,fd=5))                                     
udp   UNCONN   0        0                  0.0.0.0:7427           0.0.0.0:*      users:(("xtra_heartbeat_",pid=21390,fd=6))                                     
udp   UNCONN   0        0              192.0.23.15:53375          0.0.0.0:*      users:(("xtra_heartbeat_",pid=21390,fd=7))                                     

user@drnfff358c2: ~$ sudo ss -tulpn | grep micro_ros
udp    UNCONN   0        0               127.0.0.1:39947          0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=16))                                     
udp    UNCONN   0        0                 0.0.0.0:7448           0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=13))                                     
udp    UNCONN   0        0                 0.0.0.0:7449           0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=14))                                     
udp    UNCONN   0        0                 0.0.0.0:7450           0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=24))                                     
udp    UNCONN   0        0                 0.0.0.0:7451           0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=25))                                     
udp    UNCONN   0        0               127.0.0.1:54169          0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=27))                                     
udp    UNCONN   0        0                 0.0.0.0:46726          0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=15))                                     
udp    UNCONN   0        0                 0.0.0.0:34955          0.0.0.0:*      users:(("micro_ros_agent",pid=4303,fd=26))                                     
```

**Comment by omertal88** [comment]
My observations so far are (based on the `finest` log and capture:
1. As said before, Micro-ROS-agent uses 4 ports (2 for discovery and 2 for user data). It seems that only one of them (`10f40ab:cf10b9d1:1000000`) expired during the disconnection (as seen in line 4091 in the log).
2. Then, the expired participant sent a DATA message and `prune_deleted_participant_guid` was printed on line 4700 (couldn't find any information about what this message actually means).
3. I tried setting LeaseDuration to `90 s` but from what I could tell it didn't solve the problem (disconnection time was shorter than that)

**Comment by omertal88** [comment]
I now re-inserted my bridge-node with a different configuration and it seems to behave better now.
Haven't been able to reproduce yet.

**Comment by eboasson** [comment]
I had a look at your "finest" trace, and in that case I would say that your analysis is spot-on 😀 All I can do is provide a bit more information.

The problem lies in the interaction with the Micro-ROS agent ("MRA"), and I would expect this not to fail if the other side were Cyclone. From your earlier comment about the bridge (point 7), that seems not to be supported by your observations though.

`grep -E`'ing on `SPDP.*NEW|expired|prune|DATA.*:1000000:100c2|10f40ab:cf10b9d1:1000000:3c2` (with a few bits filtered out) we get the following scenario with regards to discovering writers on the MRA (other cases are analogous).

1. Initial discovery: SPDP arrives, triggers creation of a new proxy participant for the MRA, creation of endpoint discovery readers, receipt of a `HEARTBEAT` from MRA for advertising writer discovery data, request for retransmit:
    ```
    1678466495.033725 [0] dq.builtin: SPDP ST0 10f40ab:cf10b9d1:1000000:1c1 bes c3f NEW (data udp/192.0.23.35:7449@3 meta udp/192.0.23.35:7448@3) QOS={property_list={1:"PARTICIPANT_TYPE":"SIMPLE"}:{}}
    1678466495.047899 [0]       recv: HEARTBEAT(#41:1..28  L(:1c1 720.928893)10f40ab:cf10b9d1:1000000:3c2 -> 3f201001:efcf46e8:7690bdb5:3c7: end-of-tl-seq(rd 3f201001:efcf46e8:7690bdb5:3c7 #28) 3f201001:efcf46e8:7690bdb5:3c7@0)
    1678466495.048009 [0]        tev: acknack 3f201001:efcf46e8:7690bdb5:3c7 -> 10f40ab:cf10b9d1:1000000:3c2: F#1:1/28:1111111111111111111111111111
    1678466495.048049 [0]        tev: send acknack(rd 3f201001:efcf46e8:7690bdb5:3c7 -> pwr 10f40ab:cf10b9d1:1000000:3c2)
    ```
2. Receipt of requested data, processing of discovery of application writers on MRA:
    ```
    1678466495.075281 [0]       recv: DATA(10f40ab:cf10b9d1:1000000:3c2 -> 3f201001:efcf46e8:7690bdb5:3c7 #1 L(:1c1 720.956345))
   ...
    1678466495.076143 [0]       recv: DATA(10f40ab:cf10b9d1:1000000:3c2 -> 3f201001:efcf46e8:7690bdb5:3c7 #28 msr_in_sync(3f201001:efcf46e8:7690bdb5:3c7 out-of-sync to tlcatchup))
    1678466495.076278 [0] dq.builtin: data(builtin, vendor 1.15): 10f40ab:cf10b9d1:1000000:3c2 #1: ST0 DCPSPublication/PublicationBuiltinTopicData:{topic_name="rt/drone/heartbeat",type_name="std_msgs::msg::dds_::String_",durability=0,durability_service=0:0:1:-1:-1:-1,deadline=9223372036854775807,latency_budget=0,liveliness=0:9223372036854775807,reliability=0:100000000,lifespan=9223372036854775807,destination_order=0,ownership=0,presentation=0:0:0,time_based_filter=0,protocol_version=2:3,vendorid=1:15,participant_guid={10f40ab:cf10b9d1:1000000:1c1},keyhash={010f40ab:cf10b9d1:01000000:00000103},endpoint_guid={10f40ab:cf10b9d1:1000000:103},unicast_locator={udp/192.0.23.35:7449}}
    ```
3. Another `HEARTBEAT`, now the response is a simple ACK because all data was received.
    ```
    1678466496.054604 [0]       recv: HEARTBEAT(#42:1..28  L(:1c1 721.935689)10f40ab:cf10b9d1:1000000:3c2 -> 0:0:0:3c7: 3f201001:efcf46e8:7690bdb5:3c7@28(sync))
    1678466496.054672 [0]        tev: acknack 3f201001:efcf46e8:7690bdb5:3c7 -> 10f40ab:cf10b9d1:1000000:3c2: F#2:29/0:
    1678466496.054694 [0]        tev: send acknack(rd 3f201001:efcf46e8:7690bdb5:3c7 -> pwr 10f40ab:cf10b9d1:1000000:3c2)
    ```
4. Eventually, the lease expiry and the clean-up:
   ```
   1678466549.721361 [0]         gc: lease expired: l 0x7f740258c0 guid 10f40ab:cf10b9d1:1000000:1c1 tend 765602353184 < now 765602463328
   1678466549.726032 [0]         gc: delete_proxy_writer (10f40ab:cf10b9d1:1000000:3c2) - deleting
   1678466549.731676 [0]         gc: gc_delete_proxy_writer_dqueue(0x7f7c00fb10, 10f40ab:cf10b9d1:1000000:3c2)
   1678466549.731750 [0] dq.builtin: gc_delete_proxy_writer_dqueue_bubble(0x7f7c00fb10, 10f40ab:cf10b9d1:1000000:3c2)
   1678466549.733804 [0]         gc: gc_delete_proxy_writer(0x7f7c00fb10, 10f40ab:cf10b9d1:1000000:3c2)
   ```
5. Some time later a new SPDP message arrives from the MRA. Cyclone holds on to GUIDs of deleted participants for a while, the `prune_deleted_participant` here is simply that it throws that record away here and creates a new proxy participant.
    ```
    1678466551.175553 [0]       recv: DATA(10f40ab:cf10b9d1:1000000:100c2 -> 0:0:0:100c7 #1)
    1678466551.176116 [0] dq.builtin: prune_deleted_participant_guid(10f40ab:cf10b9d1:1000000:1c1)
    1678466551.176314 [0] dq.builtin: SPDP ST0 10f40ab:cf10b9d1:1000000:1c1 bes c3f NEW (data udp/192.0.23.35:7449@3 meta udp/192.0.23.35:7448@3) QOS={property_list={1:"PARTICIPANT_TYPE":"SIMPLE"}:{}}
    ```
6. Once again receipt of a `HEARTBEAT` (see above), and again request for retransmitting the data describing the writers (I has to, it deleted all this information on the lease expiry, so now it has to rediscover). However, there is no response ...
    ```
    1678466552.301928 [0]       recv: HEARTBEAT(#44:1..28  L(:1c1 778.182967)10f40ab:cf10b9d1:1000000:3c2 -> 0:0:0:3c7: end-of-tl-seq(rd 3f201001:efcf46e8:7690bdb5:3c7 #28) 3f201001:efcf46e8:7690bdb5:3c7@0)
    1678466552.302128 [0]        tev: acknack 3f201001:efcf46e8:7690bdb5:3c7 -> 10f40ab:cf10b9d1:1000000:3c2: F#3:1/28:1111111111111111111111111111
    1678466552.302163 [0]        tev: send acknack(rd 3f201001:efcf46e8:7690bdb5:3c7 -> pwr 10f40ab:cf10b9d1:1000000:3c2)
    1678466553.276232 [0]       recv: HEARTBEAT(#45:1..28  L(:1c1 779.157300)10f40ab:cf10b9d1:1000000:3c2 -> 0:0:0:3c7: 3f201001:efcf46e8:7690bdb5:3c7@0)
    1678466553.276353 [0]        tev: acknack 3f201001:efcf46e8:7690bdb5:3c7 -> 10f40ab:cf10b9d1:1000000:3c2: F#4:1/28:1111111111111111111111111111
    1678466553.276390 [0]        tev: send acknack(rd 3f201001:efcf46e8:7690bdb5:3c7 -> pwr 10f40ab:cf10b9d1:1000000:3c2)
    ```

Because the MRA never sends the writer discovery data that is requested, Cyclone never discovers the remote writers (and analogous cases) and things do not work. If I were to hazard a guess, it is that:
1. the MRA never observed a disconnect but Cyclone did
2. the MRA sent the `HEARTBEAT` message in response to a "pre-emptive" ACK from Cyclone
3. the MRA ignored the NACK for data that had been ACK'd in the past

Point 1 can't be helped; point 2 is perfectly acceptable (and the reason Cyclone sends these "pre-emptive" ACKs); point 3 is where it gets interesting. I'd say it is a bug in the MRA, but the spec only says it "should" retransmit in a case like this ...

**Comment by omertal88** [comment]
Thank you for your detailed answer.
I feel like I need to read this 10 more times in order to somewhat understand it.
But that's enough to get me started.. :)

**Comment by eboasson** [comment]
In DDS the participants advertise after how long a time of silence they are to be considered dead, and the `LeaseDuration` setting (unless overridden by the `Liveliness` QoS on the participant) configures this.

The way I read your question, I think you expected this to configure after how much time it will consider a remote participant, but it is the other way around: it configures for how long the remote participant should consider *this* participant alive even when it receives nothing.

**Comment by omertal88** [comment]
Oh yes you're right. I thought it was the other way around.
